### PR TITLE
Patterns from composable helpers post

### PR DESCRIPTION
**Pull request by martndemus**

### CODE REVIEW

**Comment by brzpegasus** [comment]
Is this a WIP?

**Comment by bcardarella** [comment]
@martndemus is this ready yet?

**Comment by martndemus** [comment]
@ce-doh it's now ready for review!

File: source/posts/patterns-from-composable-helpers.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+---
+layout: post
+title: "Patterns from Ember Composable Helpers"
+social: true
+author: Marten Schilstra
+twitter: "martndemus"
+github: martndemus
+summary: "A walkthrough of one of the patterns used in Ember Composable Helpers"
+published: true
+tags: Ember, JavaScript
```
Comment (bcardarella) [inline source/posts/patterns-from-composable-helpers.md]: add Engineering

File: source/posts/patterns-from-composable-helpers.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  }),
+
+  contentDidChange: observer('content', function() {
+    this.recompute();
+  })
+});
+```
+
+Let's start with the `compute` function on lines 12-17. 
+It expects a `targetObject` and `propertyPath` param, which stand for the object you wan't to get the given property from.
```
Comment (bcardarella) [inline source/posts/patterns-from-composable-helpers.md]: stands

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+tags: Ember, JavaScript, Engineering
+---
+
+Co-creator of Ember Composable Helpers [Lauren Tan][lauren] recently wrote about the [what, why and how][blogpost-lauren] of Ember Composable Helpers.
+Now I'd like to talk about one of the patterns we have used to make Ember Composable Helpers work.
+
+## The fundament of the array helpers is the get helper
+
+Most of the array helpers are built upon the implementation of the [get helper][get-helper]. 
+Let's take a closer look at a simplified version of the get helper:
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: back ticks around `get`

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+```js
+import Ember from 'ember';
+
+const { 
+  Helper, 
+  get,
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: You forgot `set` 😛

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    defineProperty(this, 'content', oneWay(`targetObject.${propertyPath}`));
+  }),
+
+  contentDidChange: observer('content', function() {
+    this.recompute();
+  })
+});
+```
+
+Let's start with the `compute` function on lines 12-17.
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: Once you add `set`, don't forget to update these line numbers

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  }),
+
+  contentDidChange: observer('content', function() {
+    this.recompute();
+  })
+});
+```
+
+Let's start with the `compute` function on lines 12-17. 
+It expects a `targetObject` and `propertyPath` param, which stands for the object you wan't to get the given property from.
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: `want` without apostrophe

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+export function getHelper([targetObject, propertyPath]) {
+  return get(targetObject, propertyPath);
+}
+
+export default helper(getHelper);
+```
+
+## Solution: Observers
+
+Yes you heard it right, observers are a perfect candidate to solve the problem that our helper won't recompute when we want it to. So lets take a look at the `propertyPathDidChange` and `contentDidChange` observers.
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: `lets` should be `let's`

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+```js
+propertyPathDidChange: observer('propertyPath', function() {
+  let propertyPath = get(this, 'propertyPath');
+  defineProperty(this, 'content', oneWay(`targetObject.${propertyPath}`));
+})
+```
+
+Let me explain what happens with this observer. On the first line we define an observer that will be triggered every time `propertyPath` gets updated. In the function body we get the value of `propertyPath` and use it to define a new computed property _at runtime_, we do that using [`defineProperty`][defineproperty]. This means that every time the `propertyPath`'s value changes, the `content` computed property gets redefined to point towards the correct path on the target object.
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: Maybe add a period after `runtime` and `We do that...` can be a separate sentence

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+contentDidChange: observer('content', function() {
+  this.recompute();
+})
+```
+
+Then there is the `contentDidChange` observer. This one watches for changes of the `content` property, which we define with the `propertyPathDidChange` observer. The `contentDidChange` observer calls `recompute`, which recomputes the end value of the helper.
+
+## Putting it all together to create the map-by helper
+
+Now that we know how to build an helper that can recompute when a property that we only know of at runtime changes, it is very simple to create other similar helpers upon this pattern. I'll leave you with the `map-by` helper, which doesn't look that different from the `get` helper i've shown you.
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: how to build `a` helper

Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: capitalize `I've`

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+```js
+import Ember from 'ember';
+
+const { 
+  Helper, 
+  get,
```
Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: `set`!

Comment (rsocci) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: Not sure why the comment showed up again at the top, but `set` here as well

File: source/posts/2016-05-13-patterns-from-composable-helpers.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+export function getHelper([targetObject, propertyPath]) {
+  return get(targetObject, propertyPath);
+}
+
+export default helper(getHelper);
+```
+
+## Solution: Observers
+
+Yes you heard it right, observers are a perfect candidate to solve the problem that our helper won't recompute when we want it to. So let's take a look at the `propertyPathDidChange` and `contentDidChange` observers.
```
Comment (poteto) [inline source/posts/2016-05-13-patterns-from-composable-helpers.md]: Maybe explain why we need to use an observer? I'm sure that will be a common question for people reading this

**Comment by bcardarella** [comment]
Is this ready to go?

**Pull request merged**